chain/db.go

```diff
@@ -181,6 +198,135 @@ func NewMemDB() *MemDB {
 	}
 }
 
+type cacheBucket struct {
+	mb memBucket
+	db DBBucket
+}
+
+func (b cacheBucket) Get(key []byte) []byte {
+	if val := b.mb.Get(key); val != nil {
+		return val
+	} else if _, ok := b.mb.db.dels[b.mb.name][string(key)]; ok {
+		return nil
+	}
+	return b.db.Get(key)
+}
+
+func (b cacheBucket) Put(key, value []byte) error {
+	err := b.mb.Put(key, value)
+	return err
+}
+
+func (b cacheBucket) Delete(key []byte) error {
+	return b.mb.Delete(key)
+}
+
+func (b cacheBucket) Iter() iter.Seq2[[]byte, []byte] {
+	return b.db.Iter() // NOTE: imperfect
+}
+
+// A CacheDB caches DB writes in memory and sorts them before flushing to the
+// underlying DB. This can greatly improve performance for certain databases.
+type CacheDB struct {
+	mem *MemDB
+	db  DB
+	kvs map[string][][2][]byte
+}
+
+// Bucket implements DB.
+func (db *CacheDB) Bucket(name []byte) DBBucket {
+	b := db.db.Bucket(name)
+	if b == nil {
+		return nil
+	} else if db.mem.Bucket(name) == nil {
+		db.mem.CreateBucket(name)
+	}
+	return cacheBucket{memBucket{string(name), db.mem}, b}
+}
+
+// CreateBucket implements DB.
+func (db *CacheDB) CreateBucket(name []byte) (DBBucket, error) {
+	if _, err := db.db.CreateBucket(name); err != nil {
+		return nil, err
+	}
+	return db.mem.CreateBucket(name)
+}
+
+// Flush implements DB.
+func (db *CacheDB) Flush() error {
+	// puts
+	for name, puts := range db.mem.puts {
+		bucket := db.kvs[name]
+		for key, val := range puts {
+			bucket = append(bucket, [2][]byte{[]byte(key), val})
+		}
+		db.kvs[name] = bucket
+	}
+	for bucket, kvs := range db.kvs {
+		bucket := db.db.Bucket([]byte(bucket))
+		sort.Slice(kvs, func(i, j int) bool {
+			return bytes.Compare(kvs[i][0], kvs[j][0]) < 0
+		})
+		for _, kv := range kvs {
+			bucket.Put(kv[0], kv[1])
```

Comment:
Might be overlooking something but is there a reason we don't process `dels` to potentially save a db write here?